csi/moac/test/watcher_test.js

```diff
@@ -36,490 +58,460 @@ const Readable = require('stream').Readable;
-// in order to be able to feed data to it etc.
-class StreamMockTracker {
-  constructor () {
-    this.current = null;
-  }
-
-  // create a new stream (mimics nodejs k8s client api)
-  getObjectStream () {
-    const s = new StreamMock();
-    this.current = s;
-    return s;
-  }
-
-  // get the most recently created underlaying stream
-  latest () {
-    return this.current;
-  }
+// Create a cache with a listWatch object with fake start method that does
+// nothing instead of connecting to k8s cluster.
+function createMockedCache () {
+  const kc = new KubeConfig();
+  Object.assign(kc, fakeConfig);
+  const watcher = new CustomResourceCache('namespace', 'apple', kc, Apple, {
+    restartDelay: RESTART_DELAY_MS,
+    eventTimeout: EVENT_TIMEOUT_MS,
+    idleTimeout: IDLE_TIMEOUT_MS
+  });
+  // convenience function for generating k8s watcher events
+  watcher.emitKubeEvent = (ev, data) => {
+    watcher.listWatch.callbackCache[ev].forEach((cb) => cb(data));
+  };
+  const startStub = sinon.stub(watcher.listWatch, 'start');
+  startStub.onCall(0).resolves();
+  return [watcher, startStub];
 }
 
 module.exports = function () {
-  // Basic watcher operations grouped in describe to avoid repeating watcher
-  // initialization & tear down for each test case.
-  describe('watch events', () => {
-    var getMock = new GetMock();
-    var streamMock = new StreamMock();
-    var watcher;
-    var newList = [];
-    var modList = [];
-    var delList = [];
-
-    before(() => {
-      watcher = new Watcher('test', getMock.template(), streamMock, objectFilter);
-      watcher.on('new', (obj) => newList.push(obj));
-      watcher.on('mod', (obj) => modList.push(obj));
-      watcher.on('del', (obj) => delList.push(obj));
-
-      getMock.add(createObject('valid-object', 1, 123));
-      getMock.add(createObject('invalid-object', 1, 99));
-    });
-
-    after(() => {
-      watcher.stop();
-      streamMock.end();
+  this.timeout(10000);
+
+  it('should create a cache and block in start until connected', async () => {
+    const kc = new KubeConfig();
+    Object.assign(kc, fakeConfig);
+    const watcher = new CustomResourceCache('namespace', 'apple', kc, Apple, {
+      restartDelay: RESTART_DELAY_MS,
+      eventTimeout: EVENT_TIMEOUT_MS
     });
+    const startStub = sinon.stub(watcher.listWatch, 'start');
+    startStub.onCall(0).rejects();
+    startStub.onCall(1).rejects();
+    startStub.onCall(2).resolves();
+    const startTime = new Date();
+    await watcher.start();
+    const delta = new Date() - startTime;
+    sinon.assert.calledThrice(startStub);
+    expect(watcher.isConnected()).to.be.true();
+    expect(delta).to.be.within(2 * RESTART_DELAY_MS, 3 * RESTART_DELAY_MS);
+    watcher.stop();
+  });
 
-    it('should init cache only with objects which pass through the filter', async () => {
-      await watcher.start();
+  it('should reconnect watcher if it gets disconnected', async () => {
```

Comment:
if we get disconnected are we able to get the events we missed while disconnected? I'm guessing we do because the watcher will get the latest once connected, and if something is not on the cache or is now gone we'll get the appropriate event?

Reply:
There is a resource version parameter in watcher URL and the API server will start sending all events that are more recent than the specified resource version. So the watcher gets all objects and then starts watcher with resource version obtained from the object list. Then we compare the new list with what we had in the cache before and generate appropriate add, del, mod events. The only thing to be aware of is that the mod event is emitted even for objects that have not changed. So a consumer of these events must handle gracefully that case. All our operators do so, so not a problem for us.